src/sync/generic_network_transport.hpp

```diff
@@ -0,0 +1,354 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+
+#ifndef REALM_GENERIC_NETWORK_TRANSPORT_HPP
+#define REALM_GENERIC_NETWORK_TRANSPORT_HPP
+
+#include <functional>
+#include <string>
+#include <memory>
+#include <map>
+#include <vector>
+
+#include <realm/util/to_string.hpp>
+
+namespace realm {
+namespace app {
+namespace error {
+
+#pragma mark Errors
+
+struct AppError : public std::runtime_error
+{
+    enum class Type {
+        Unknown,
+        JSON,
+        Service,
+        Custom
+    };
+
+    const std::string category() const
```

Comment:
Returning `const std::string` doesn't do anything useful; it just forces the string to be copied an extra time.